### ISSUE TITLE
Demangle pytorch and tensorflow dependencies

### ISSUE THREAD

**Issue by CloseChoice** [shap/shap#3662]
### Issue Description

Currently our dependencies for the deep explainer are mangled together somewhere deep down. This best visible when running the notebook `notebooks/image_examples/image_classification/PyTorch Deep Explainer MNIST example.ipynb` without `tf_keras` installed, then one gets an error. This is completely counter intuitive since this notebook just needs pytorch related dependencies. We should definitely get this into a state where pytorch and tensorflow code can coexist without the other framework being installed.

### Minimal Reproducible Example

```python
pip uninstall tf_keras

then run `notebooks/image_examples/image_classification/PyTorch Deep Explainer MNIST example.ipynb`

then this throws an error in the cell



# since shuffle=True, this is a random sample of test data
batch = next(iter(test_loader))
images, _ = batch

background = images[:100]
test_images = images[100:103]

e = shap.DeepExplainer(model, background)
shap_values = e.shap_values(test_images)
``
```

### Traceback

```shell
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File ~/programming/github/shap_master/.venv/lib/python3.10/site-packages/transformers/activations_tf.py:22
     21 try:
---> 22     import tf_keras as keras
     23 except (ModuleNotFoundError, ImportError):

ModuleNotFoundError: No module named 'tf_keras'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
File ~/programming/github/shap_master/.venv/lib/python3.10/site-packages/transformers/utils/import_utils.py:1472, in _LazyModule._get_module(self, module_name)
   1471 try:
-> 1472     return importlib.import_module("." + module_name, self.__name__)
   1473 except Exception as e:

File /usr/lib/python3.10/importlib/__init__.py:126, in import_module(name, package)
    125         level += 1
--> 126 return _bootstrap._gcd_import(name[level:], package, level)

File <frozen importlib._bootstrap>:1050, in _gcd_import(name, package, level)

File <frozen importlib._bootstrap>:1027, in _find_and_load(name, import_)
...
   1476         f" traceback):\n{e}"
   1477     ) from e

RuntimeError: Failed to import transformers.modeling_tf_utils because of the following error (look up to see its traceback):
Your currently installed version of Keras is Keras 3, but this is not yet supported in Transformers. Please install the backwards-compatible tf-keras package with `pip install tf-keras`.
```

### Expected Behavior

should work without any tensorflow dependencies

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

master

**Comment by Anushreebasics** [comment]
Hi @CloseChoice , I would like to solve this issue

**Comment by ashwinkothavade** [comment]
Hi @CloseChoice — I'd like to take this on.

Before filing a PR I'd start by reproducing in a clean torch-only env (no `tensorflow` or `tf_keras` installed) and tracing which import chain is actually forcing TF. A quick look at the code suggests a few things are already right:

- `shap/explainers/_deep/__init__.py` already lazy-imports `deep_tf` vs `deep_pytorch` based on the detected framework, so `DeepExplainer` on a `torch.nn.Module` shouldn't pull `shap/explainers/_deep/deep_tf.py`.
- `shap/explainers/_gradient.py` and `shap/explainers/tf_utils.py` already use function-local TF imports in most places.

So the leak is probably somewhere else on the `import shap.DeepExplainer` path — either an eager import in `shap/explainers/__init__.py` or a helper module (`shap/models/_text_generation.py`, `_topk_lm.py`, `_teacher_forcing.py` all mention keras/tf-adjacent code). I'll find the exact offender by reproducing, then scope the PR to that.

Plan for the first PR:
1. Add a CI test that installs **torch only** (no tensorflow, no tf_keras) and runs a minimal PyTorch `DeepExplainer` example. Currently there's no guardrail that catches this regression.
2. Move the offending top-level TF imports to function-local / `try/except`-guarded imports.
3. Keep scope to the `_deep` + direct-dependency import chain; if more modules need the same treatment I'd follow up with a part 2.

A couple of questions before I start:
- Is the failing notebook the `pytorch_mnist` one under `notebooks/api_examples/explainers/`? Want to make sure I'm reproducing the exact flow you had in mind.
- Are you OK with the CI addition running in a separate, minimal-deps job (since the existing test matrix has both frameworks installed)?

Targeting a draft PR within 4–5 days.

**Comment by Anushreebasics** [comment]
@ashwinkothavade I am already working on this, and have already raised the PR, you are welcome to review it :)

**Comment by ashwinkothavade** [comment]
> [@ashwinkothavade](https://github.com/ashwinkothavade) I am already working on this, and have already raised the PR, you are welcome to review it :)

You're right @Anushreebasics — sorry, I missed your prior claim and #4612 when I opened this. Closing.

Happy to review #4612 as a user. If a torch-only CI guard seems useful as a follow-up after your fix lands, the subprocess-based test in this branch (`test/deep-explainer-torch-only-regression` on my fork) is yours to lift verbatim or ignore — no strings attached.
